src/PointwiseFunctions/GeneralRelativity/WeylTypeD1.cpp

```diff
@@ -0,0 +1,172 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/GeneralRelativity/WeylTypeD1.hpp"
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/VectorImpl.hpp"
+#include "PointwiseFunctions/GeneralRelativity/WeylElectric.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+//
+
+namespace {
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::Ij<DataType, SpatialDim, Frame> up_down_from_lower_rank_2_tensor(
+    const tnsr::ii<DataType, SpatialDim, Frame>& lower_rank_2_tensor,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric) {
+  auto result = make_with_value<tnsr::Ij<DataType, SpatialDim, Frame>>(
+      get<0, 0>(inverse_spatial_metric), 0.0);
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    for (size_t j = 0; j < SpatialDim; ++j) {
+      for (size_t k = 0; k < SpatialDim; ++k) {
+        result.get(j, k) +=
+            lower_rank_2_tensor.get(i, k) * inverse_spatial_metric.get(i, j);
+      }
+    }
+  }
+  return result;
+}
+
+}  // namespace
+
+namespace gr {
+template <size_t SpatialDim, typename Frame, typename DataType>
+void weyl_type_D1(
+    const gsl::not_null<tnsr::ii<DataType, SpatialDim, Frame>*> weyl_type_D1,
+    const tnsr::ii<DataType, SpatialDim, Frame>& weyl_electric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric) {
+  auto weyl_electric_up_down =
+      up_down_from_lower_rank_2_tensor(weyl_electric, inverse_spatial_metric);
+
+  auto upper_weyl_electric =
+      make_with_value<tnsr::II<DataType, SpatialDim, Frame>>(
+          get<0, 0>(inverse_spatial_metric), 0.0);
```

Comment:
[optional] This function does a lot of memory allocations.  Since it looks like this is mostly for observing, that's not a big deal, but you will get better performance if you use a `Variables` with temporary tags, as is done for `Psi4`, for example.